yellow-paper/docs/l1-smart-contracts/frontier.md

```diff
@@ -0,0 +1,158 @@
+---
+title: Frontier Merkle Tree
+---
+
+The Frontier Merkle Tree is an append only Merkle tree that is optimized for minimal storage on chain.
+By storing only the right-most non-empty node at each level of the tree we can always extend the tree with a new leaf or compute the root without needing to store the entire tree.
+We call these values the frontier of the tree.
+If we have the next index to insert at and the current frontier, we have everything we need to extend the tree or compute the root, with much less storage than a full merkle tree.
+
+We will go through a few diagrams and explanations to understand how this works.
+And then a pseudo implementation is provided.
+
+
+## Insertion
+Whenever we are inserting, we need to update the "root" of the largest subtree possible.
+This is done by updating the node at the level of the tree, where we have just inserted its right-most descendant.
+This can sound a bit confusing, so we will go through a few examples. 
+
+At first, say that we have the following tree, and that it is currently entirely empty.
+
+![alt text](images/proposal/image-1.png)
```

Comment:
Maybe move these to `imagees/frontier/image-n.png`?

Reply:
Will fix